### PR TITLE
Distinguish between simple matches with and without the terms index

### PR DESCRIPTION
**Pull request by romseygeek**
We currently use TextSearchInfo to let query parsers know when a field
will support match queries.  Some field types (numeric, constant, range)
can produce simple match queries that don't use the terms index, and
it is useful to distinguish between these fields on the one hand and 
keyword/text-type fields on the other.  In particular, the 
SignificantTextAggregation only works on fields that have indexed terms,
but there is currently no efficient way to see this at search time and so
the factory falls back on checking to see if an index analyzer has been
defined, with the result that some nonsensical field types are permitted.

This commit adds a new static TextSearchInfo implementation called
`SIMPLE_MATCH_WITHOUT_TERMS` that can be returned by field
types with no corresponding terms index.  It changes significant text
to check for this rather than for the presence of an index analyzer.

This is a breaking change, in that the significant text agg will now throw 
an error up-front if you try and apply it to a numeric field, whereas before 
you would get an empty result.

### CODE REVIEW

**Review by polyfractal — APPROVED** [review]
This is great!  I stumbled into that oddity when working up tests for SigText after the VS refactoring, but didn't really have an idea how to fix it cleanly at the time :)

Probably best to wait on Mark's review since he knows the agg better, but :+1: from me

Do we need to add a note to the breaking-changes doc page?

File: server/src/main/java/org/elasticsearch/index/mapper/VersionFieldMapper.java [hunk 1]
```diff
@@ -42,7 +42,7 @@
         public static final VersionFieldType INSTANCE = new VersionFieldType();
 
         private VersionFieldType() {
-            super(NAME, false, false, true, TextSearchInfo.SIMPLE_MATCH_ONLY, Collections.emptyMap());
+            super(NAME, false, false, true, TextSearchInfo.NONE, Collections.emptyMap());
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/VersionFieldMapper.java:45]: This seems to have slipped when I double checked all the mappers, maybe it does not have a test so we don't enforce its consistency.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/SignificantTextAggregatorFactory.java [hunk 2]
```diff
@@ -98,6 +99,13 @@ public SignificantTextAggregatorFactory(String name,
         this.significanceHeuristic = significanceHeuristic;
     }
 
+    private static boolean supportsAgg(MappedFieldType ft) {
+        if (ft.getTextSearchInfo() == TextSearchInfo.NONE) {
+            return false;
+        }
+        return ft.getTextSearchInfo() != TextSearchInfo.SIMPLE_MATCH_WITHOUT_TERMS;
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/SignificantTextAggregatorFactory.java:106]: should it be return ft.getTextSearchInfo != NONE && gt.getTextSearchIngo != SIMPLE_MATCH_WITHOUT_TERMS ?

**Review by javanna — APPROVED** [review]
I left a couple of minor comments/questions, LGTM otherwise

File: server/src/internalClusterTest/java/org/elasticsearch/search/aggregations/bucket/SignificantTermsSignificanceScoreIT.java [hunk 3]
```diff
@@ -590,11 +578,7 @@ public void testScriptCaching() throws Exception {
                 .getMissCount(), equalTo(1L));
 
         // Ensure that non-scripted requests are cached as normal
-        if (useSigText) {
-            r = client().prepareSearch("cache_test_idx").setSize(0).addAggregation(significantText("foo", "s")).get();
-        } else {
-            r = client().prepareSearch("cache_test_idx").setSize(0).addAggregation(significantTerms("foo").field("s")).get();
-        }
+        r = client().prepareSearch("cache_test_idx").setSize(0).addAggregation(significantTerms("foo").field("s")).get();
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/internalClusterTest/java/org/elasticsearch/search/aggregations/bucket/SignificantTermsSignificanceScoreIT.java:581]: Hm, you're right, I had thought this was just testing a useless agg but actually I should change it instead to also have some text and to run the sigText agg over the text field instead.

File: server/src/main/java/org/elasticsearch/index/mapper/VersionFieldMapper.java (hunk 1 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/VersionFieldMapper.java:45]: Yes, there's no test for VersionFieldMapper - I'll add one as a follow up.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/SignificantTextAggregatorFactory.java (hunk 2 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/SignificantTextAggregatorFactory.java:106]: ++

**Comment by romseygeek** [comment]
> Do we need to add a note to the breaking-changes doc page?

We do, but I think I can only do that on backport as we don't have per-7.x-version changes in master?

File: server/src/main/java/org/elasticsearch/index/mapper/VersionFieldMapper.java (hunk 1 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/VersionFieldMapper.java:45]: thank you

File: server/src/internalClusterTest/java/org/elasticsearch/search/aggregations/bucket/SignificantTermsSignificanceScoreIT.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/internalClusterTest/java/org/elasticsearch/search/aggregations/bucket/SignificantTermsSignificanceScoreIT.java]: ++

**Comment by markharwood** [comment]
Looks good but I'm not sure about the `terms` part of naming `TextSearchInfo.SIMPLE_MATCH_WITHOUT_TERMS`.
While it relates to Lucene's `termsEnum` it may be misinterpreted as relating to supporting `term` queries.

Number fields support a `terms` query e.g. this works:

    "term":{"num_tags" : "1"}

I recall they actually do a range query behind the scenes.
I don't have a better suggestion for `SIMPLE_MATCH_WITHOUT_TERMS` currently.

**Review by markharwood — COMMENTED** [review]
Is there a test that checks annotated_text is supported by significant_text?

**Comment by polyfractal** [comment]
> We do, but I think I can only do that on backport as we don't have per-7.x-version changes in master? 

Doh, yes, ignore me.  :)

**Comment by romseygeek** [comment]
> it may be misinterpreted as relating to supporting term queries.

I'm less worried about this, because it's an internal name and I think the javadoc is clear enough?

> Is there a test that checks annotated_text is supported by significant_text?

There isn't, I'll see if I can add one.

**Comment by romseygeek** [comment]
Adding aggs tests to annotated text turns out to not be trivial.  In the interest of not blocking this PR, I've opened a separate [issue](https://github.com/elastic/elasticsearch/issues/64202) to deal with that.

**Comment by romseygeek** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-unix (gradle download failure)

**Pull request merged**